lib/src/system_cache.dart

```diff
@@ -146,4 +141,72 @@ class SystemCache {
     log.fine('Clean up system cache temp directory $tempDir.');
     if (dirExists(tempDir)) deleteEntry(tempDir);
   }
+
+  /// An in-memory cache of pubspecs described by [describe].
+  final cachedPubspecs = <PackageId, Pubspec>{};
+
+  /// Loads the (possibly remote) pubspec for the package version identified by
+  /// [id].
+  ///
+  /// This may be called for packages that have not yet been downloaded during
+  /// the version resolution process. Its results are automatically memoized.
+  ///
+  /// Throws a [DataException] if the pubspec's version doesn't match [id]'s
+  /// version.
+  Future<Pubspec> describe(PackageId id) async {
```

Comment:
It seems weird to me that these things live on the `SystemCache`.
In my mind it makes more sense that a `Source` instance would be passed `SystemCache` when created, and then registered with `SourceRegistry`. And then `SourceRegistry` is responsible for calling methods on `Source` and ideally nobody else would ever interact with concrete source implementations.

Everything goes through `SourceRegistry`, and the rest of the code can call methods here... and when it returns a `PackageRef` / `PackageRange` / `PackageId` then the `Description` / `ResolvedDescription` are private members on  `PackageRef` / `PackageRange` / `PackageId`. But `PackageRef` / `PackageRange` / `PackageId` allows for calling generic methods available on `Description` (like serialize for `pubspec.yaml` or `pubspec.lock`, etc).

But I'm guessing we might have leaked the abstractions too many places to clean this up now, right?

I'm just asking questions here, not even suggesting we write a TODO, hehe :D

Reply:
I'm not 100% sure we couldn't avoid leaking the abstractions here. But if it's even remotely possible it's probably best to leave it for a follow-up PR.